### PR TITLE
Fix polling on https connections in HttpConnectionPool

### LINKED ISSUE #49431: HttpConnection regression around read-aheads on pooled https connections
When we take a connection out of the SocketsHttpHandler connection pool to process a request, we want to validate that it hasn't received any stray data, so we need to poll it.  But we also know we're about to issue a read on it as part of handling the request.  So, we just issue an async read as part of taking it out of the pool, and check whether that read completes, to avoid the extra poll sys call.  However, the other time we poll connections is while they're in the pool, every once in a while checking every connection to see if it's still valid, but not removing it from the pool.  Here, we don't want to do such a read-ahead, for several reasons:
1. On Windows, doing an async read with a buffer pins that buffer for the duration of the read.  That means we'd end up pinning a buffer for every connection in the pool and for as long as that connection is idle in the pool, which has negative impact on GC.
2. If there's a pending read on a connection we then tear down, it results in an exception, which is both extra overhead and has caused developers to complain about unexpected unobserved task exceptions.  We've fixed bugs in the past related to this.

So, when we create connections, even as we wrap the socket in layers of streams, we try to hold on to the socket so we can Poll it in the case where we want to check connection viability but without taking it out of the pool.  In .NET Core 3.1, we were able to do this for most connections, both http and https, and only couldn't when using a proxy tunnel.
https://github.com/dotnet/corefx/blob/5e7fb3dbb33257d285857e3f9f654df457234173/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L646

But as part of adding (https://github.com/dotnet/runtime/commit/4b8c5fe2b4757f77f5699d49a072fcb123d2e33b#diff-bc2faad6c76ecb9b115e305763eb55ebfde74de1eee27ed549e5edaabb097a99R1180) and then reverting (https://github.com/dotnet/runtime/commit/3609f05a6631e4a533392eff5753ad08506600c1#diff-bc2faad6c76ecb9b115e305763eb55ebfde74de1eee27ed549e5edaabb097a99R1271), this Socket-tracking code was (presumably accidentally) removed for .NET 5, with it now only checking for NetworkStream:
https://github.com/dotnet/runtime/commit/3609f05a6631e4a533392eff5753ad08506600c1#diff-595f700a4c83b735d856fe14b5ebe8c7f7c18ae6e586adc8b398a997a0678274R82-R85
This means that whereas in .NET Core 3.1 we would only poll rather than read-ahead for https connections when periodically checking their status in the pool, now in .NET 5 we're doing read-aheads for https connections, and incurring the aforementioned problems that come with it (buffers pinned for long periods, extra exceptions when connections are torn down).

### PR DESCRIPTION
**Pull request by stephentoub**
Fixes https://github.com/dotnet/runtime/issues/49431

Avoid performing read-aheads on such connections, which can result in long-pinned buffers and unnecessary exceptions when the connections are torn down.

cc: @geoffkizer, @scalablecory

### CODE REVIEW

**Review by scalablecory — APPROVED** [review]

**Comment by scalablecory** [comment]
Actually, before you merge this:

TLS can send alerts out of band from stream data right @wfurt? Presumably this "check that we haven't received anything" would fail with this change.

**Review by avparuch — APPROVED** [review]

**Review by wfurt — APPROVED** [review]
LGTM. I'm wondering if long-term we can use the 0-read @geoffkizer recently fixed. It is somewhat sad that using Stream has negative impact.

**Comment by wfurt** [comment]
> TLS can send alerts out of band from stream data right @wfurt? Presumably this "check that we haven't received anything" would fail with this change.

If the alert is coming from remote side, then yes. It can come at any time. That can be close notification or the peer did not like something we sent. Typically that also terminates the connection shortly after. We can also generate alert on our side as result of error condition.

My assumption this generally moves back to state we already had before NetworkStream.Socket was available, right @stephentoub ?

**Comment by stephentoub** [comment]
> My assumption this generally moves back to state we already had before NetworkStream.Socket was available

This moves us back to the state we were in for .NET 3.1: all http and https connections, except for those via proxy tunnels, poll the underlying socket for connections in the pool.  With this change, we'll now also do it if a NetworkStream is returned from ConnectCallback (which obviously wasn't available in 3.1).

**Comment by stephentoub** [comment]
Even if such an alert happens, it seems like the worst case here is a race condition where we try to use a connection that's about to be closed and then we end up needing to retry... but such a race condition already exists,  which is why we already have said retry code.  Yes?

**Comment by scalablecory** [comment]
If the only thing an alert can be used for is an error condition then lets merge away 👍. Do we know that TLS 1.3 is the same @wfurt?

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs [hunk 1]
```diff
@@ -1393,10 +1404,15 @@ private async ValueTask<Stream> ApplyPlaintextFilterAsync(bool async, Stream str
...
 
-        private async ValueTask<HttpConnection> ConstructHttp11ConnectionAsync(bool async, Stream stream, TransportContext? transportContext, HttpRequestMessage request, CancellationToken cancellationToken)
+        private async ValueTask<HttpConnection> ConstructHttp11ConnectionAsync(bool async, Socket? socket, Stream stream, TransportContext? transportContext, HttpRequestMessage request, CancellationToken cancellationToken)
         {
-            stream = await ApplyPlaintextFilterAsync(async, stream, HttpVersion.Version11, request, cancellationToken).ConfigureAwait(false);
-            return new HttpConnection(this, stream, transportContext);
+            Stream newStream = await ApplyPlaintextFilterAsync(async, stream, HttpVersion.Version11, request, cancellationToken).ConfigureAwait(false);
+            if (newStream != stream)
+            {
+                // If a plaintext filter created a new stream, we can't trust that the socket is still applicable, so rely on it.
```
Comment (scalablecory) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:1412]:
```suggestion
                // If a plaintext filter created a new stream, we can't trust that the socket is still applicable.
```

**Comment by wfurt** [comment]
> If the only thing an alert can be used for is an error condition then lets merge away 👍. Do we know that TLS 1.3 is the same @wfurt?

yes in regard to alerts. The biggest changes were in TLS handshake.

**Pull request merged**